README.md
# Next.js Bundle Analysis Github Action

Analyzes each PR's impact on your next.js app's bundle size and displays it using a comment. Optionally supports performance budgets.

![screenshot of bundle analysis comment](https://p176.p0.n0.cdn.getcloudapp.com/items/BluKP76d/2b51f74a-9c0f-481f-b76a-9b36cf37d369.png?v=ddd23d0d9ee1ee9ad40487d181ed917f)

## Installation

It's pretty simple to get this set up, just run the following command and answer the prompts. The command will create a `.github/workflows` directory in your project root and add a `next_bundle_analysis.yml` file to it - that's all it takes!

```sh
$ npx -p nextjs-bundle-analysis generate
```

> **NOTE**: Due to github actions' lack of support for more complex actions, the experience of getting this set up is unusual in that it requires a generation script which copies most of the logic into your project directly. As soon as github adds support for the [features](https://github.com/actions/runner/pull/1144) [needed](https://github.com/actions/runner/pull/1144#discussion_r651087316) to properly package up this action, we'll put out an update that removes the extra boilerplate and makes usage much simpler. Until then, we all have no choice but to endure this unusual setup process.

## Configuration

Config values are written to `package.json` under the key `nextBundleAnalysis`, and can be changed there any time. You can directly edit the workflow file if you want to adjust your default branch or the directory that your nextjs app lives in (especially if you are using a `srcDir` or something similar).

### `showDetails (boolean)`

(Optional, defaults to `true`) This option renders a collapsed "details" section under each section of the bundle analysis comment explaining some of the finer details of the numbers provided. If you feel like this is not necessary and you and/or those working on your project understand the details, you can set this option to `false` and that section will not render.

### `buildOutputDirectory (string)`

(Optional, defaults to `.next`) If your application [builds to a custom directory](https://nextjs.org/docs/api-reference/next.config.js/setting-a-custom-build-directory), you can specify this with the key `buildOutputDirectory`. You will also need to replace all instances of `.next` in `next_bundle_analysis.yml` with your custom output directory.

For example, if you build to `dist`, you should:

- Set `package.json.nextBundleAnalysis.buildOutputDirectory` to `"dist"`.
- In `nextjs_bundle_analysis`, replace all instances of `.next` with `dist`.

### `budget (number)`

(Optional) The file size, in bytes, to budget for first page load size. For example, if `budget` was set to `358400` (350 KB) and a page's first load size was 248 KB, the report would list that page as having used 70% of the performance budget.

### `budgetPercentIncreaseRed (number)`

(Optional, but required if `budget` is specified) If a page's first load size has increased more than `budgetPercentIncreaseRed` percent, display a 🔴 to draw attention to the change.

### `minimumChangeThreshold (number)`

(Optional, defaults to `0`) The threshold under which pages will be considered unchanged. For example, if `minimumChangeThreshold` was set to `500` and a page's size increased by `300 B`, it will be considered unchanged.

### `skipCommentIfEmpty (boolean)`

(Optional, defaults to `false`) When set to `true`, if no pages have changed size the generated comment will be an empty string.

## Caveats

- This plugin only analyzes the direct bundle output from next.js. If you have added any other scripts via the `<script>` tag, especially third party scripts and things like analytics or other tracking scripts, these are not included in the analysis. Scripts of this nature should _probably_ be loaded in behind a consent manager and should not make an impact on your initial load, and as long as this is how you handle them it should make no difference, but it's important to be aware of this and account for the extra size added by these scripts if they are present in your app.
- Since this plugin works by comparing the base bundle against each PR, the first time it is run, it will fail since it has no base to compare against. This is expected - ideally you can just commit the changes directly to your default branch, where it will run to generate the base bundle, then anything that branches off after that will have a valid comparison point and the script will work as expected.
- This script assumes that running `next build` will successfully build your application. If you need additional scripts or logic to do so, you may need to update the action step called "Build next.js app" to the command needed to build your app. For example, if you have a `npm run build` step, that would be a good target to change it to. We plan to make this configurable via the generator in the future.


CHANGELOG.md
# nextjs-bundle-analysis

## 0.5.0

### Minor Changes

- [#44](https://github.com/hashicorp/nextjs-bundle-analysis/pull/44) [`067cb5d`](https://github.com/hashicorp/nextjs-bundle-analysis/commit/067cb5de682d51ad30b8aac2f69437303825c3b8) Thanks [@dstaley](https://github.com/dstaley)! - Improve monorepo support by adding the `name` from `package.json` to the generated comment. Add support for a `skipCommentIfEmpty` configuration option that will set the comment to an empty string when no pages have changed size.


__tests__/index.js
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

const path = require('path')
const fs = require('fs')
const { execSync } = require('child_process')
const rimraf = require('rimraf')
const { afterEach, beforeEach, expect } = require('@jest/globals')
const mkdirp = require('mkdirp')
const { getBuildOutputDirectory, getOptions } = require('../utils')

const fixturesPath = path.join(__dirname, '__fixtures__')

// Get all test suites (fixtures)
const fixtures = fs
  .readdirSync(fixturesPath, { withFileTypes: true })
  .filter((dirent) => dirent.isDirectory())
  .map((dirent) => dirent.name)

describe('sort of integration', () => {
  fixtures.forEach((dirName) => {
    describe(`fixture ${dirName}`, () => {
      const cwd = path.join(fixturesPath, dirName)
      const options = getOptions(cwd)
      const buildOutputDirectory = getBuildOutputDirectory(options)

      beforeEach(() => {
        process.chdir(cwd)
        execSync('npm install')
        execSync('npm run build')
      })

      afterEach(() => {
        rimraf.sync(path.join(cwd, buildOutputDirectory))
      })

      test(`bundle analysis action generates report and compares artifacts correctly ${dirName}`, () => {
        // make sure the 'report' command works
        execSync('node ../../../report.js')
        const bundleAnalysis = fs.readFileSync(
          path.join(
            process.cwd(),
            buildOutputDirectory,
            'analyze/__bundle_analysis.json'
          ),
          'utf8'
        )
        expect(bundleAnalysis.length).toBeGreaterThan(1)

        // create a fake artifact download - in the real world this would pull from
        // github as part of the action flow
        mkdirp.sync(
          path.join(process.cwd(), buildOutputDirectory, 'analyze/base/bundle')
        )
        fs.writeFileSync(
          path.join(
            process.cwd(),
            buildOutputDirectory,
            'analyze/base/bundle/__bundle_analysis.json'
          ),
          bundleAnalysis
        )

        // make sure the 'compare' command works
        execSync('node ../../../compare.js')
        const comment = fs.readFileSync(
          path.join(
            process.cwd(),
            buildOutputDirectory,
            'analyze/__bundle_analysis_comment.txt'
          ),
          'utf8'
        )
        expect(comment).toMatch(/no changes to the javascript bundle/)
      })
    })
  })
})


__tests__/__fixtures__/custom-output-directory/next.config.js
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

module.exports = {
  distDir: 'dist',
}


report.js
#!/usr/bin/env node
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */


const path = require('path')
const fs = require('fs')
const gzSize = require('gzip-size')
const mkdirp = require('mkdirp')
const { getBuildOutputDirectory, getOptions } = require('./utils')

// Pull options from `package.json`
const options = getOptions()
const BUILD_OUTPUT_DIRECTORY = getBuildOutputDirectory(options)

// first we check to make sure that the build output directory exists
const nextMetaRoot = path.join(process.cwd(), BUILD_OUTPUT_DIRECTORY)
try {
  fs.accessSync(nextMetaRoot, fs.constants.R_OK)
} catch (err) {
  console.error(
    `No build output found at "${nextMetaRoot}" - you may not have your working directory set correctly, or not have run "next build".`
  )
  process.exit(1)
}

// if so, we can import the build manifest
const buildMeta = require(path.join(nextMetaRoot, 'build-manifest.json'))

// this memory cache ensures we dont read any script file more than once
// bundles are often shared between pages
const memoryCache = {}

// since _app is the template that all other pages are rendered into,
// every page must load its scripts. we'll measure its size here
const globalBundle = buildMeta.pages['/_app']
const globalBundleSizes = getScriptSizes(globalBundle)

// next, we calculate the size of each page's scripts, after
// subtracting out the global scripts
const allPageSizes = Object.values(buildMeta.pages).reduce(
  (acc, scriptPaths, i) => {
    const pagePath = Object.keys(buildMeta.pages)[i]
    const scriptSizes = getScriptSizes(
      scriptPaths.filter((scriptPath) => !globalBundle.includes(scriptPath))
    )

    acc[pagePath] = scriptSizes
    return acc
  },
  {}
)

// format and write the output
const rawData = JSON.stringify({
  ...allPageSizes,
  __global: globalBundleSizes,
})

// log ouputs to the gh actions panel
console.log(rawData)

mkdirp.sync(path.join(nextMetaRoot, 'analyze/'))
fs.writeFileSync(
  path.join(nextMetaRoot, 'analyze/__bundle_analysis.json'),
  rawData
)

// --------------
// Util Functions
// --------------

// given an array of scripts, return the total of their combined file sizes
function getScriptSizes(scriptPaths) {
  const res = scriptPaths.reduce(
    (acc, scriptPath) => {
      const [rawSize, gzipSize] = getScriptSize(scriptPath)
      acc.raw += rawSize
      acc.gzip += gzipSize
      return acc
    },
    { raw: 0, gzip: 0 }
  )
  return res
}

// given an individual path to a script, return its file size
function getScriptSize(scriptPath) {
  const encoding = 'utf8'
  const p = path.join(nextMetaRoot, scriptPath)

  let rawSize, gzipSize
  if (Object.keys(memoryCache).includes(p)) {
    rawSize = memoryCache[p][0]
    gzipSize = memoryCache[p][1]
  } else {
    const textContent = fs.readFileSync(p, encoding)
    rawSize = Buffer.byteLength(textContent, encoding)
    gzipSize = gzSize.sync(textContent)
    memoryCache[p] = [rawSize, gzipSize]
  }

  return [rawSize, gzipSize]
}


compare.js
#!/usr/bin/env node
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

const originalFilesize = require('filesize')
const numberToWords = require('number-to-words')
const fs = require('fs')
const path = require('path')
const { getBuildOutputDirectory, getOptions } = require('./utils')

// Override default filesize options to display a non-breakable space as a spacer.
const filesize = (bytes, options) => {
  return originalFilesize(bytes, {
    spacer: ' ',
    ...options,
  })
}

// Pull options from `package.json`
const options = getOptions()

const BUDGET = options.budget
const BUDGET_PERCENT_INCREASE_RED = options.budgetPercentIncreaseRed
// this must be explicitly set to false not to render
const SHOW_DETAILS =
  options.showDetails === undefined ? true : options.showDetails
const BUILD_OUTPUT_DIRECTORY = getBuildOutputDirectory(options)
const PACKAGE_NAME = options.name
const SKIP_COMMENT_IF_EMPTY = options.skipCommentIfEmpty

// import the current and base branch bundle stats
const currentBundle = require(path.join(
  process.cwd(),
  BUILD_OUTPUT_DIRECTORY,
  'analyze/__bundle_analysis.json'
))
const baseBundle = require(path.join(
  process.cwd(),
  BUILD_OUTPUT_DIRECTORY,
  'analyze/base/bundle/__bundle_analysis.json'
))

// kick it off
let output = `## 📦 Next.js Bundle Analysis for ${PACKAGE_NAME}

This analysis was generated by the [Next.js Bundle Analysis action](https://github.com/hashicorp/nextjs-bundle-analysis). 🤖

`

// pull the global bundle out, we handle this separately
const globalBundleCurrent = currentBundle.__global
const globalBundleBase = baseBundle.__global
delete currentBundle.__global
delete baseBundle.__global

// calculate the difference between the current bundle and the base branch's
let globalBundleChanges = false
const globalGzipDiff = globalBundleCurrent.gzip - globalBundleBase.gzip
// only report a global bundle size change if we don't have a minimum change
// threshold configured, or if the change is greater than the threshold
if (
  globalGzipDiff !== 0 &&
  (!('minimumChangeThreshold' in options) ||
    Math.abs(globalGzipDiff) > options.minimumChangeThreshold)
) {
  globalBundleChanges = {
    page: 'global',
    raw: globalBundleCurrent.raw,
    gzip: globalBundleCurrent.gzip,
    gzipDiff: globalGzipDiff,
    increase: Math.sign(globalGzipDiff) > 0,
  }
}

// now we're going to go through each of the pages in the current bundle and
// run analysis on each one.
const changedPages = []
const newPages = []

for (let page in currentBundle) {
  const currentStats = currentBundle[page]
  const baseStats = baseBundle[page]

  // if the page does't appear in the base bundle, it is a new page, we can
  // push this directly to its own category. we also don't compare it to anything
  // because its a new page.
  if (!baseStats) {
    newPages.push({ page, ...currentStats })
  } else if (currentStats.gzip !== baseStats.gzip) {
    // otherwise, we run a comparsion between the current page and base branch page
    // we push these to their own category for rendering later
    const rawDiff = currentStats.raw - baseStats.raw
    const gzipDiff = currentStats.gzip - baseStats.gzip
    const increase = !!Math.sign(gzipDiff)
    // only report a page size change if we don't have a minimum change
    // threshold configured, or if the change is greater than the threshold
    if (
      !('minimumChangeThreshold' in options) ||
      Math.abs(gzipDiff) > options.minimumChangeThreshold
    ) {
      changedPages.push({ page, ...currentStats, rawDiff, gzipDiff, increase })
    }
  }
}

// with our data in hand, we now get to a bunch of output formatting.
// we start with any changes to the global bundle.
if (globalBundleChanges) {
  // start with the headline, which will render differently depending on whether
  // there was an increase of decrease.
  output += `### ${
    globalBundleChanges.increase ? '⚠️' : '🎉'
  }  Global Bundle Size ${
    globalBundleChanges.increase ? 'Increased' : 'Decreased'
  }
  
`
  // this is where we actually generate the table including the changes.
  output += markdownTable(globalBundleChanges)

  // and we end with some extra details further explaining the data above
  if (SHOW_DETAILS) {
    output += `\n<details>
<summary>Details</summary>
<p>The <strong>global bundle</strong> is the javascript bundle that loads alongside every page. It is in its own category because its impact is much higher - an increase to its size means that every page on your website loads slower, and a decrease means every page loads faster.</p>
<p>Any third party scripts you have added directly to your app using the <code>&lt;script&gt;</code> tag are not accounted for in this analysis</p>
<p>If you want further insight into what is behind the changes, give <a href='https://www.npmjs.com/package/@next/bundle-analyzer'>@next/bundle-analyzer</a> a try!</p>
</details>\n\n`
  }
}

// next up is the newly added pages
if (newPages.length) {
  // this might seem like too much, but I feel like this type of small detail really
  // matters <3
  const plural = newPages.length > 1 ? 's' : ''
  output += `### New Page${plural} Added
  
The following page${plural} ${
    plural === 's' ? 'were' : 'was'
  } added to the bundle from the code in this PR:

`
  // as before, run the data in as a table
  output += markdownTable(newPages, globalBundleCurrent) + '\n'

  // there is no "details" section here, didnt't seem necessary. i'm open to one being
  // added though!
}

// finally, we run through the pages that existed in the base branch, still exist in the
// current branch, and have changed size.
if (changedPages.length) {
  // same flow here as the others:
  // - headline that adjusts wording based on number of changes
  // - table containing all the resources and info
  // - details section
  const plural = changedPages.length > 1 ? 's' : ''
  output += `### ${titleCase(
    numberToWords.toWords(changedPages.length)
  )} Page${plural} Changed Size
  
The following page${plural} changed size from the code in this PR compared to its base branch:

`
  output += markdownTable(changedPages, globalBundleCurrent, globalBundleBase)

  // this details section is a bit more responsive, it will render slightly different
  // details depending on whether a budget is being used, since the information presented
  // is quite different.
  if (SHOW_DETAILS) {
    output += `\n<details>
<summary>Details</summary>
<p>Only the gzipped size is provided here based on <a href='https://twitter.com/slightlylate/status/1412851269211811845'>an expert tip</a>.</p>
<p><strong>First Load</strong> is the size of the global bundle plus the bundle for the individual page. If a user were to show up to your website and land on a given page, the first load size represents the amount of javascript that user would need to download. If <code>next/link</code> is used, subsequent page loads would only need to download that page's bundle (the number in the "Size" column), since the global bundle has already been downloaded.</p>
<p>Any third party scripts you have added directly to your app using the <code>&lt;script&gt;</code> tag are not accounted for in this analysis</p>
${
  BUDGET && globalBundleCurrent
    ? `<p>The "Budget %" column shows what percentage of your performance budget the <strong>First Load</strong> total takes up. For example, if your budget was 100kb, and a given page's first load size was 10kb, it would be 10% of your budget. You can also see how much this has increased or decreased compared to the base branch of your PR. If this percentage has increased by ${BUDGET_PERCENT_INCREASE_RED}% or more, there will be a red status indicator applied, indicating that special attention should be given to this. If you see "+/- <0.01%" it means that there was a change in bundle size, but it is a trivial enough amount that it can be ignored.</p>`
    : `<p>Next to the size is how much the size has increased or decreased compared with the base branch of this PR. If this percentage has increased by ${BUDGET_PERCENT_INCREASE_RED}% or more, there will be a red status indicator applied, indicating that special attention should be given to this.`
}
</details>\n`
  }
}

// and finally, if there are no changes at all, we try to be clear about that
const hasNoChanges =
  !newPages.length && !changedPages.length && !globalBundleChanges
if (hasNoChanges) {
  output += 'This PR introduced no changes to the JavaScript bundle! 🙌'
}

// however, if ignoreIfEmpty is true, set output to an empty string
if (hasNoChanges && SKIP_COMMENT_IF_EMPTY) {
  output = ''
}

// log the output, mostly for testing and debugging. this will show up in the
// github actions console.
console.log(output)

// and to cap it off, we write the output to a file which is later read in as comment
// contents by the actions workflow.
fs.writeFileSync(
  path.join(
    process.cwd(),
    BUILD_OUTPUT_DIRECTORY,
    'analyze/__bundle_analysis_comment.txt'
  ),
  output.trim()
)

// Util Functions

// this is where the vast majority of the complexity lives, its a single function
// that renders a markdown table displaying a wide range of bundle size data in a
// wide variety of different ways. this could potentially be improved by splitting it
// up into several different functions for rendering the different tables we produce
// (new pages, changed pages, global bundle)
function markdownTable(_data, globalBundleCurrent, globalBundleBase) {
  const data = [].concat(_data)
  // the table renders different depending on whether the budget option is enabled
  // and also some tables do not run budget diffs (new, global)
  const showBudget = globalBundleCurrent && BUDGET
  const showBudgetDiff = BUDGET && !!globalBundleBase

  // first we set up the table headers
  return `Page | Size (compressed) | ${
    globalBundleCurrent ? `First Load |` : ''
  }${showBudget ? ` % of Budget (\`${filesize(BUDGET)}\`) |` : ''}
|---|---|${globalBundleCurrent ? '---|' : ''}${showBudget ? '---|' : ''}
${data
  .map((d) => {
    // next, we go through each item in the bundle data that was passed in and render
    // a row for it. a couple calculations are run upfront to make rendering easier.
    const firstLoadSize = globalBundleCurrent
      ? d.gzip + globalBundleCurrent.gzip
      : 0

    const budgetPercentage = showBudget
      ? ((firstLoadSize / BUDGET) * 100).toFixed(2)
      : 0
    const previousBudgetPercentage =
      globalBundleBase && d.gzipDiff
        ? (
            ((globalBundleCurrent.gzip + d.gzip + d.gzipDiff) / BUDGET) *
            100
          ).toFixed(2)
        : 0
    const budgetChange = previousBudgetPercentage
      ? (previousBudgetPercentage - budgetPercentage).toFixed(2)
      : 0

    return (
      `| \`${d.page}\`` +
      renderSize(d, showBudgetDiff) +
      renderFirstLoad(globalBundleCurrent, firstLoadSize) +
      renderBudgetPercentage(
        showBudget,
        budgetPercentage,
        previousBudgetPercentage,
        budgetChange
      ) +
      ' |\n'
    )
  })
  .join('')}`
}

// as long as global bundle is passed, render the first load size, which is the global
// bundle plus the size of the current page, representing the total JS required
// in order to land on that page.
function renderFirstLoad(globalBundleCurrent, firstLoadSize) {
  if (!globalBundleCurrent) return ''
  return ` | ${filesize(firstLoadSize)}`
}

// renders the bundle size of the current page. if there is a diff from the base branch
// size of the page, also displays the size difference, unless there is a budget set and
// the budget has a diff from the base branch, in which case the diff is not rendered.
function renderSize(d, showBudgetDiff) {
  const gzd = d.gzipDiff
  const percentChange = (gzd / d.gzip) * 100
  return ` | \`${filesize(d.gzip)}\`${
    gzd && !showBudgetDiff
      ? ` _(${renderStatusIndicator(percentChange)}${filesize(gzd)})_`
      : ''
  }`
}

// renders the percentage of the budget taken up by the current page's first load js
// for changed pages, also renders the percent change compared to the base branch size
function renderBudgetPercentage(
  showBudget,
  budgetPercentage,
  previousBudgetPercentage,
  budgetChange
) {
  if (!showBudget) return ''

  // we round to 2 decimal places for number values, if there was a change smaller than that
  // its displayed as "+/- <0.01%", signaling that it's not a consequential change, but it
  // still is a change technically, so we still show it.
  const budgetChangeText = ` _(${renderStatusIndicator(budgetChange)}${
    budgetChange < 0.01 && budgetChange > -0.01
      ? '+/- <0.01%'
      : budgetChange + '%'
  })_`

  // only render the budget change for changed pages (indicated by previousBudgetPercentage
  // being passed in)
  return ` | ${budgetPercentage}%${
    previousBudgetPercentage ? budgetChangeText : ''
  }`
}

// given a percentage that a metric has changed, renders a colored status indicator
// this makes it easier to call attention to things that need attention
//
// in general:
// - yellow means "keep an eye on this"
// - red means "this is a problem"
// - green means "this is a win"
function renderStatusIndicator(percentageChange) {
  let res = ''
  if (percentageChange > 0 && percentageChange < BUDGET_PERCENT_INCREASE_RED) {
    res += '🟡 +'
  } else if (percentageChange >= BUDGET_PERCENT_INCREASE_RED) {
    res += '🔴 +'
  } else if (percentageChange < 0.01 && percentageChange > -0.01) {
    res += ''
  } else {
    res += '🟢 '
  }
  return res
}

function titleCase(str) {
  return str.replace(/\w\S*/g, (txt) => {
    return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()
  })
}


generate.mjs
#!/usr/bin/env node
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */
import {
  intro,
  outro,
  confirm,
  select,
  spinner,
  isCancel,
  cancel,
  text,
  group,
  note,
} from '@clack/prompts'
import color from 'picocolors'
import path from 'node:path'
import fs from 'node:fs'
import mkdirp from 'mkdirp'

const WORKFLOW_TEMPLATE_FILE = 'template.yml'
const WORKFLOW_FILE = 'next_bundle_analysis.yml'

const DEFAULT_PACKAGE_CONFIG = {
  budget: 350,
  budgetPercentIncreaseRed: 20,
  minimumChangeThreshold: 0,
  buildOutputDirectory: '.next',
  showDetails: true,
}

const DEFAULT_WORKFLOW_CONFIG = {
  baseBranch: 'main',
  nodeVersion: 18,
  packageManager: 'npm',
  workingDirectory: './',
  buildCommand: './node_modules/.bin/next build',
  buildOutputDirectory: DEFAULT_PACKAGE_CONFIG.buildOutputDirectory,
}

async function number(opts) {
  const result = await text(opts)
  return Number.parseInt(result, 10)
}

async function writePackageJsonConfig(config) {
  // write the config values to package.json
  const packageJsonPath = path.join(process.cwd(), 'package.json')
  const packageJsonContent = JSON.parse(
    fs.readFileSync(packageJsonPath, 'utf-8')
  )
  packageJsonContent.nextBundleAnalysis = {
    ...DEFAULT_PACKAGE_CONFIG,
    ...config,
  }
  fs.writeFileSync(packageJsonPath, JSON.stringify(packageJsonContent, null, 2))
}

async function writeWorkflowFile(config) {
  const packageJsonPath = new URL('package.json', import.meta.url)
  const packageJsonContent = JSON.parse(
    fs.readFileSync(packageJsonPath, 'utf-8')
  )
  const templatePath = new URL(WORKFLOW_TEMPLATE_FILE, import.meta.url)
  const workflowsPath = path.join(process.cwd(), '.github/workflows')
  const workflowFilePath = path.join(workflowsPath, WORKFLOW_FILE)

  let template = fs.readFileSync(templatePath, 'utf-8')

  // Specify the latest version
  template = template.replace('{PACKAGE_VERSION}', packageJsonContent.version)

  // mkdir -p the .workflows directory
  mkdirp.sync(workflowsPath)

  const areInputsChangedFromDefault = !Object.keys(
    DEFAULT_WORKFLOW_CONFIG
  ).every((key) => DEFAULT_WORKFLOW_CONFIG[key] === config[key])

  if (areInputsChangedFromDefault) {
    // update the base branch
    template = template.replace('- main', `- ${config.baseBranch}`)

    // Update the inputs
    template = template
      .replace('# with:', 'with:')
      .replace(/#\s+node-version:.+$/m, `  node-version: ${config.nodeVersion}`)
      .replace(
        /#\s+package-manager:.+$/m,
        `  package-manager: ${config.packageManager}`
      )
      .replace(
        /#\s+working-directory:.+$/m,
        `  working-directory: ${config.workingDirectory}`
      )
      .replace(
        /#\s+build-output-directory:.+$/m,
        `  build-output-directory: ${config.buildOutputDirectory}`
      )
      .replace(
        /#\s+build-command:.+$/m,
        `  build-command: ${config.buildCommand}`
      )
  }

  fs.writeFileSync(workflowFilePath, template)
}

async function main() {
  console.log('\n', color.inverse(color.bold(' nextjs-bundle-analysis ')), '\n')

  intro(color.inverse(' configuration '))

  const packageConfig = await group({
    budget: async () => {
      const setBudget = await confirm({
        message: 'Would you like to set a performance budget?',
      })

      if (setBudget) {
        return (
          (await number({
            message: `What would you like the maximum javascript on first load to be (in kb)? (default: ${DEFAULT_PACKAGE_CONFIG.budget})`,
            defaultValue: DEFAULT_PACKAGE_CONFIG.budget,
          })) * 1024
        )
      }

      return DEFAULT_PACKAGE_CONFIG.budget
    },
    budgetPercentIncreaseRed: () =>
      number({
        message: `If you exceed this percentage of the budget or filesize, it will be highlighted in red (default: ${DEFAULT_PACKAGE_CONFIG.budgetPercentIncreaseRed})`,
        defaultValue: DEFAULT_PACKAGE_CONFIG.budgetPercentIncreaseRed,
      }),
    minimumChangeThreshold: () =>
      number({
        message: `If a page's size change is below this threshold (in bytes), it will be considered unchanged (default: ${DEFAULT_PACKAGE_CONFIG.minimumChangeThreshold})`,
        defaultValue: DEFAULT_PACKAGE_CONFIG.minimumChangeThreshold,
      }),
    buildOutputDirectory: () =>
      text({
        message: `Do you have a custom dist directory? (default: ${DEFAULT_PACKAGE_CONFIG.buildOutputDirectory})`,
        defaultValue: DEFAULT_PACKAGE_CONFIG.buildOutputDirectory,
      }),
  })

  await writePackageJsonConfig(packageConfig)

  outro('✅ Bundle analysis config written to package.json')

  intro(color.inverse(' workflow file '))

  const workflowConfig = await group({
    baseBranch: () =>
      text({
        message: `What's your base branch? (default: ${DEFAULT_WORKFLOW_CONFIG.baseBranch})`,
        defaultValue: DEFAULT_WORKFLOW_CONFIG.baseBranch,
      }),
    nodeVersion: () =>
      text({
        message: `What node version are you using? (default: ${DEFAULT_WORKFLOW_CONFIG.nodeVersion})`,
        defaultValue: DEFAULT_WORKFLOW_CONFIG.nodeVersion,
      }),
    packageManager: () =>
      select({
        message: 'What package manager do you use?',
        options: [
          { value: 'npm', label: 'npm', hint: 'default' },
          { value: 'yarn', label: 'yarn' },
          { value: 'pnpm', label: 'pnpm' },
        ],
        defaultValue: DEFAULT_WORKFLOW_CONFIG.nodeVersion,
      }),
    workingDirectory: () =>
      text({
        message: `What directory does your app live in? (default: ${DEFAULT_WORKFLOW_CONFIG.workingDirectory})`,
        defaultValue: DEFAULT_WORKFLOW_CONFIG.workingDirectory,
      }),
    buildCommand: () =>
      text({
        message: "What's your build command? (default: next build)",
        defaultValue: DEFAULT_WORKFLOW_CONFIG.buildCommand,
      }),
  })

  await writeWorkflowFile({
    ...workflowConfig,
    buildOutputDirectory: packageConfig.buildOutputDirectory,
  })

  outro(
    '✅ Workflow file written to .github/workflows/next-js-bundle-analysis.yml'
  )
}

main()


generate.js
#!/usr/bin/env node
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */


const path = require('path')
const fs = require('fs')
const mkdirp = require('mkdirp')
const inquirer = require('inquirer')

inquirer
  .prompt([
    {
      type: 'confirm',
      name: 'useBudget',
      message: 'Would you like to set a performance budget?',
      default: true,
    },
    {
      type: 'number',
      name: 'budget',
      message:
        'What would you like the maximum javascript on first load to be (in kb)?',
      default: 350,
      when: (answers) => answers.useBudget,
    },
    {
      type: 'number',
      name: 'redIndicatorPercentage',
      message:
        'If you exceed this percentage of the budget or filesize, it will be highlighted in red',
      default: 20,
    },
    {
      type: 'number',
      name: 'minimumChangeThreshold',
      message: `If a page's size change is below this threshold (in bytes), it will be considered unchanged'`,
      default: 0,
    },
  ])
  .then((answers) => {
    // write the config values to package.json
    const packageJsonPath = path.join(process.cwd(), 'package.json')
    const packageJsonContent = require(packageJsonPath)
    packageJsonContent.nextBundleAnalysis = {
      budget: answers.budget * 1024,
      budgetPercentIncreaseRed: answers.redIndicatorPercentage,
      minimumChangeThreshold: answers.minimumChangeThreshold,
      showDetails: true, // add a default "showDetails" argument
    }
    fs.writeFileSync(
      packageJsonPath,
      JSON.stringify(packageJsonContent, null, 2)
    )
    // mkdir -p the .workflows directory
    const workflowsPath = path.join(process.cwd(), '.github/workflows')
    mkdirp.sync(workflowsPath)

    // copy the template to it
    const templatePath = path.join(__dirname, 'template.yml')
    const destinationPath = path.join(
      workflowsPath,
      'nextjs_bundle_analysis.yml'
    )
    fs.copyFileSync(templatePath, destinationPath)
  })
